edgedb/datatypes/namedtuple.c

```diff
@@ -338,35 +351,103 @@ static PyMethodDef namedtuple_methods[] = {
 };
 
 
-static PySequenceMethods namedtuple_as_sequence = {
-    .sq_length = (lenfunc)namedtuple_length,
-    .sq_item = (ssizeargfunc)namedtuple_getitem,
+static PyType_Slot namedtuple_slots[] = {
+    {Py_tp_repr, (reprfunc)namedtuple_repr},
+    {Py_tp_methods, namedtuple_methods},
+    {Py_tp_getattro, (getattrofunc)namedtuple_getattr},
+    {Py_tp_traverse, (traverseproc)namedtuple_traverse},
+    {Py_tp_dealloc, (destructor)namedtuple_dealloc},
+    {Py_tp_new, namedtuple_derived_new},
+    {0, 0},
+};
+
+
+static PyType_Spec namedtuple_spec = {
+    "edgedb.DerivedNamedTuple",
+    sizeof(PyTupleObject) - sizeof(PyObject *),
+    sizeof(PyObject *),
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    namedtuple_slots,
 };
 
 
 PyTypeObject EdgeNamedTuple_Type = {
     PyVarObject_HEAD_INIT(NULL, 0)
     "edgedb.NamedTuple",
-    .tp_basicsize = sizeof(EdgeNamedTupleObject) - sizeof(PyObject *),
+    .tp_basicsize = sizeof(PyTupleObject) - sizeof(PyObject *),
     .tp_itemsize = sizeof(PyObject *),
-    .tp_methods = namedtuple_methods,
-    .tp_dealloc = (destructor)namedtuple_dealloc,
-    .tp_as_sequence = &namedtuple_as_sequence,
-    .tp_hash = (hashfunc)namedtuple_hash,
-    .tp_getattro = (getattrofunc)namedtuple_getattr,
-    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
-    .tp_richcompare = (richcmpfunc)namedtuple_richcompare,
-    .tp_traverse = (traverseproc)namedtuple_traverse,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
     .tp_new = namedtuple_new,
-    .tp_free = PyObject_GC_Del,
-    .tp_repr = (reprfunc)namedtuple_repr,
-    .tp_weaklistoffset = offsetof(EdgeNamedTupleObject, weakreflist),
 };
 
 
+PyObject *
+EdgeNamedTuple_Type_New(PyObject *desc)
+{
+    assert(init_type_called);
+
+    if (desc == NULL || !EdgeRecordDesc_Check(desc)) {
+        PyErr_BadInternalCall();
+        return NULL;
+    }
+
+    PyObject *type;
+    PyObject *rv;
+
+    type = PyType_FromSpecWithBases(
+        &namedtuple_spec, PyTuple_Pack(1, &EdgeNamedTuple_Type)
+    );
+    if (type == NULL) {
+        return NULL;
+    }
+
+    // Over-allocate the new type object to store a quick pointer to desc.
+    PyObject_GC_UnTrack(type);  // needed by PyObject_GC_Resize
+    // PyObject_GC_Resize() increases the size by type->ob_type->tp_itemsize,
+    // which is sizeof(PyMemberDef) in PyType_Type, but we only need PyObject*,
+    // so make sure we have enough space to store the pointer.
+    Py_ssize_t size = Py_SIZE(type);
+    assert(Py_TYPE(type)->tp_itemsize > sizeof(PyObject *));
+    rv = PyObject_GC_Resize(PyObject, type, size  + 1);
+    if (rv == NULL) {
+        PyObject_GC_Del(type);
+        return NULL;
+    }
+    Py_SIZE(rv) = size;
+    EdgeNamedTuple_Type_DESC(rv) = desc;
+    // desc is also stored in tp_dict for refcount.
+    PyDict_SetItemString(((PyTypeObject *)rv)->tp_dict, "__desc__", desc);
```

Comment:
Because the Stable ABI doesn't have the `PyHeapTypeObject` type, we're using this hack to store a fast access for the `desc` pointer.

Reply:
It turns out that we cannot blindly resize a heap type object, because `PyType_Ready()` sets several pointers like `__mro__` and subclass weakrefs in base classes, the address of the type object is not expected to change after that.